src/Native/System.Net.Security.Native/pal_gssapi.cpp

```diff
@@ -0,0 +1,265 @@
+                                                    uint32_t statusValue,
+                                                    int32_t isGssMechCode,
+                                                    GssBuffer** outBufferHandle,
+                                                    uint32_t* statusLength)
+{
+    assert(minorStatus != nullptr);
+    assert(outBufferHandle != nullptr);
+    assert(statusLength != nullptr);
+    assert(isGssMechCode == 0 || isGssMechCode == 1);
+    int statusType = isGssMechCode ? GSS_C_MECH_CODE : GSS_C_GSS_CODE;
+    *outBufferHandle = new GssBuffer();
+    uint32_t majorStatus =
+        gss_display_status(minorStatus, statusValue, statusType, GSS_C_NO_OID, nullptr, *outBufferHandle);
+    NetSecurityNative_HandleError(majorStatus, outBufferHandle, statusLength);
+    return majorStatus;
+}
+
+extern "C" uint32_t
+NetSecurityNative_ImportUserName(uint32_t* minorStatus, char* inputName, uint32_t inputNameLen, GssName** outputName)
+{
+    assert(outputName != nullptr);
+    assert(minorStatus != nullptr);
+    assert(inputName != nullptr);
+    assert(outputName != nullptr);
+    gss_buffer_desc inputNameBuffer{.length = inputNameLen, .value = inputName};
+    gss_OID nameType = const_cast<gss_OID>(GSS_C_NT_USER_NAME);
+    return gss_import_name(minorStatus, &inputNameBuffer, nameType, outputName);
+}
+
+extern "C" uint32_t NetSecurityNative_ImportPrincipalName(uint32_t* minorStatus,
+                                                          char* inputName,
+                                                          uint32_t inputNameLen,
+                                                          GssName** outputName)
+{
+    assert(minorStatus != nullptr);
+    assert(inputNameLen >= 0);
+    assert(outputName != nullptr);
+    gss_buffer_desc inputNameBuffer{.length = inputNameLen, .value = inputName};
+    gss_OID nameType = const_cast<gss_OID>(GSS_KRB5_NT_PRINCIPAL_NAME);
+    return gss_import_name(minorStatus, &inputNameBuffer, nameType, outputName);
+}
+
+extern "C" uint32_t NetSecurityNative_InitSecContext(uint32_t* minorStatus,
+                                                     GssCredId* claimantCredHandle,
+                                                     GssCtxId** contextHandle,
+                                                     uint32_t isNtlm,
+                                                     GssName* targetName,
+                                                     uint32_t reqFlags,
+                                                     uint8_t* inputBytes,
+                                                     uint32_t inputLength,
+                                                     GssBuffer** outBufferHandle,
+                                                     uint32_t* outTokenLength,
+                                                     uint32_t* retFlags)
+{
+    assert(isNtlm == 0 || isNtlm == 1);
```

Comment:
A bunch of the arg pointers aren't verified with asserts, e.g. claimantCredHandle, targetName, inputBytes, inputLength, etc.